message/validation/partial_validation.go

```diff
@@ -0,0 +1,190 @@
+	switch msgType {
+	case spectypes.PostConsensusPartialSig,
+		spectypes.RandaoPartialSig,
+		spectypes.SelectionProofPartialSig,
+		spectypes.ContributionProofs,
+		spectypes.ValidatorRegistrationPartialSig:
+		return true
+	default:
+		return false
+	}
+}
+
+func (mv *MessageValidator) partialSignatureTypeMatchesRole(msgType spectypes.PartialSigMsgType, role spectypes.BeaconRole) bool {
+	switch role {
+	case spectypes.BNRoleAttester:
+		return msgType == spectypes.PostConsensusPartialSig
+	case spectypes.BNRoleAggregator:
+		return msgType == spectypes.PostConsensusPartialSig || msgType == spectypes.SelectionProofPartialSig
+	case spectypes.BNRoleProposer:
+		return msgType == spectypes.PostConsensusPartialSig || msgType == spectypes.RandaoPartialSig
+	case spectypes.BNRoleSyncCommittee:
+		return msgType == spectypes.PostConsensusPartialSig
+	case spectypes.BNRoleSyncCommitteeContribution:
+		return msgType == spectypes.PostConsensusPartialSig || msgType == spectypes.ContributionProofs
+	case spectypes.BNRoleValidatorRegistration:
+		return msgType == spectypes.ValidatorRegistrationPartialSig
+	default:
+		panic("invalid role") // role validity should be checked before
+	}
+}
+
+func (mv *MessageValidator) validPartialSignatures(share *ssvtypes.SSVShare, signedMsg *spectypes.SignedPartialSignatureMessage) error {
+	if err := mv.validateSignatureFormat(signedMsg.Signature); err != nil {
+		return err
+	}
+
+	if err := ssvtypes.VerifyByOperators(signedMsg.Signature, signedMsg, mv.netCfg.Domain, spectypes.PartialSignatureType, share.Committee); err != nil {
+		signErr := ErrInvalidSignature
+		signErr.innerErr = err
+		signErr.got = fmt.Sprintf("domain %v from %v", hex.EncodeToString(mv.netCfg.Domain[:]), hex.EncodeToString(share.ValidatorPubKey))
+		return signErr
+	}
+
+	seen := map[[32]byte]struct{}{}
+	for _, message := range signedMsg.Message.Messages {
+		if _, ok := seen[message.SigningRoot]; ok {
+			return fmt.Errorf("duplicated partial signature message")
+		}
+		seen[message.SigningRoot] = struct{}{}
+
+		if err := mv.validateSignatureFormat(message.PartialSignature); err != nil {
+			return err
+		}
+
+		if err := mv.verifyPartialSignature(message, share); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (mv *MessageValidator) verifyPartialSignature(msg *spectypes.PartialSignatureMessage, share *ssvtypes.SSVShare) error {
+	signer := msg.Signer
+	signature := msg.PartialSignature
+	root := msg.SigningRoot
+
+	for _, n := range share.Committee {
+		if n.GetID() != signer {
+			continue
+		}
+
+		pk, err := ssvtypes.DeserializeBLSPublicKey(n.GetPublicKey())
+		if err != nil {
+			return fmt.Errorf("deserialize pk: %w", err)
+		}
+		sig := &bls.Sign{}
+		if err := sig.Deserialize(signature); err != nil {
+			return fmt.Errorf("deserialize signature: %w", err)
+		}
+
+		if !sig.VerifyByte(&pk, root[:]) {
+			return fmt.Errorf("wrong signature")
+		}
```

Comment:
i think this can use the `BatchVerifier.AggregateVerify` method as well to increase performance (only difference is here we pass only one public key, if im not wrong)